File: src/com/train/algorithm/tree/implementInC++/SumRoottoLeafNumbers.cpp
//
// Created by 罗斯 on 2/11/2022.
//
#include <functional>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    int sumNumbers(TreeNode* root) {
        int res = 0;
        function<void(TreeNode*, int)> traverse = [&](TreeNode* t, int num){
            if (!t) return ;
            num = num * 10 + t->val;
            if (t->left || t->right){
                traverse(t->left, num);
                traverse(t->right, num);
            }else {
                res += num;
            }
        };

        traverse(root, 0);
        return res;
    }
};

File: src/com/train/algorithm/tree/implementInPython/SumRootToLeafNumbers.py
class TreeNode(object):
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution(object):
    def sumNumbers(self, root):
        def isLeaf(node):
            if node is None:
                return False
            elif node.left is None and node.right is None:
                return True
            else:
                return False
        l = []
        def dfs(node, sum):
            if node is None:
                return
            elif isLeaf(node):
                l.append(10*sum + node.val)
            dfs(node.left, 10*sum + node.val)
            dfs(node.right, 10*sum + node.val)
        dfs(root, 0)
        return sum(l)